lib/compress/zstd_compress.c

```diff
@@ -3796,6 +3801,20 @@ static size_t ZSTD_compressStream_generic(ZSTD_CStream* zcs,
             return ERROR(init_missing);
 
         case zcss_load:
+            if ( (flushMode == ZSTD_e_end)
+              && ((size_t)(oend-op) >= ZSTD_compressBound(iend-ip))
+              && (zcs->inBuffPos == 0) ) {
```

Comment:
Nope. If `inbuffPos == inToCompress`, it merely means we have no data stored within internal buffer to compress another block. But previous data can still span 2 segments, which is something we want to avoid : in such case, the compressor would automatically discard the oldest segment, keeping only the newest one in memory. This would negatively impact compression ratio.
And we want this optimization to have no impact on compression ratio (it's merely an _opportunity_).

With `inbuffPos==0`, we are guaranteed that, either there is nothing yet within internal buffer (most expected scenario), or, we have reached the top of internal round buffer and restart from the beginning, in which case past history is necessarily a single segment, which can be referenced by `ZSTD_compressEnd()`.

There is another case which can be valid, though it's not done here and is more complex to describe :
- next block must lay in a single segment (`inbuffPos == inToCompress`)
- history must be a single segment (`lowLimit == dictLimit`)

Technically, it can only happen while filling internal round buffer during the first pass (srcSize < windowSize).
After that point, history is always separated into 2 segments, except when `inbuffPos==0`.

Not sure yet if this use case is common enough to be worth implementing.